flang/lib/Lower/Bridge.cpp

```diff
@@ -801,161 +798,255 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+      genFIRIncrementLoopBegin(info);
+
+    // Generate loop body code.  The NonLabelDoStmt and EndDoStmt genFIR calls
+    // are nops, since their code is generated directly here.  However, their
+    // genFIR wrapper calls are needed for block management in some cases.
     for (auto &e : eval.getNestedEvaluations())
       genFIR(e, unstructuredContext);
-    setCurrentEval(eval);
 
-    // Generate end loop code.
+    // Generate loop end code.
     if (infiniteLoop || whileCondition) {
-      genBranch(doStmtEval.localBlocks[0]);
+      genFIRBranch(headerBlock);
     } else {
-      for (auto i = incrementLoopInfo.size(); i > 0;)
-        genFIRIncrementLoopEnd(incrementLoopInfo[--i]);
+      for (auto d = incrementLoopInfo.size(); d > 0;)
+        genFIRIncrementLoopEnd(incrementLoopInfo[--d]);
     }
   }
 
   /// Generate FIR to begin a structured or unstructured increment loop.
   void genFIRIncrementLoopBegin(IncrementLoopInfo &info) {
     auto loc = toLocation();
-    mlir::Type type =
-        info.isStructured() ? builder->getIndexType() : info.loopVariableType;
-    auto lowerValue = genFIRLoopIndex(info.lowerExpr, type);
-    auto upperValue = genFIRLoopIndex(info.upperExpr, type);
-    // clang-format off
-    info.stepValue =
-        info.stepExpr.has_value() ? genFIRLoopIndex(*info.stepExpr, type)
-        : info.isStructured()
-            ? builder->create<mlir::ConstantIndexOp>(loc, 1)
-            : builder->createIntegerConstant(loc, info.loopVariableType, 1);
-    // clang-format on
-    assert(info.stepValue && "step value must be set");
-    info.loopVariable = createTemp(loc, *info.loopVariableSym);
+    info.loopVariable = createTemp(loc, info.loopVariableSym);
+    auto controlType = info.isStructured() ? builder->getIndexType()
+                                           : genType(info.loopVariableSym);
+    auto genControlValue = [&](const Fortran::semantics::SomeExpr *expr) {
+      if (expr)
+        return builder->createConvert(loc, controlType, genExprValue(*expr));
+      if (!info.hasRealControl)
+        return builder->createIntegerConstant(loc, controlType, 1); // step
+      auto one =
+          builder->createIntegerConstant(loc, builder->getIndexType(), 1);
+      return builder->createConvert(loc, controlType, one); // real step
+    };
+    auto lowerValue = genControlValue(info.lowerExpr);
+    auto upperValue = genControlValue(info.upperExpr);
+    info.stepValue = genControlValue(info.stepExpr);
+
+    auto genLocalInitAssignments = [&]() {
+      for (const auto *sym : info.localInitSymList) {
+        llvm_unreachable("do concurrent locality specs not implemented");
+        const auto *hostDetails =
+            sym->detailsIf<Fortran::semantics::HostAssocDetails>();
+        assert(hostDetails && "missing local_init variable host variable");
+        [[maybe_unused]] const Fortran::semantics::Symbol &hostSym =
+            hostDetails->symbol();
+        // assign sym = hostSym
```

Comment:
Is this incomplete? Better to crash and burn than silently drop the baby on the floor...

Reply:
The `llvm_unreachable` call on line 932 crashes nicely for any `local_init` variable.  The call has to be inside the loop to account for the empty loop case.  I could have moved the call to just before `// assign` comment to at least exercise one loop iteration to that point, but whatever.  I did test this loop with debug output before activating the crash call.  The idea was to implement as much of the code as possible, absent symbol map support.  `local` locality variables also trigger a crash call elsewhere (earlier).  Some `local` cases would otherwise sometimes appear like they work now, but even those were really incorrect, so all of those cases crash regardless.